database-plugins/src/main/java/co/cask/hydrator/plugin/DBUtils.java

```diff
@@ -99,29 +100,50 @@ public static DriverCleanup ensureJDBCDriverIsAvailable(Class<? extends Driver>
    * Given the result set, get the metadata of the result set and return
    * list of {@link co.cask.cdap.api.data.schema.Schema.Field},
    * where name of the field is same as column name and type of the field is obtained using {@link DBUtils#getType(int)}
+   *
    * @param resultSet result set of executed query
    * @param schemaStr schema string to override resultant schema
-   *
    * @return list of schema fields
    * @throws SQLException
    * @throws IOException
    */
-  public static List<Schema.Field> getSchemaFields(ResultSet resultSet, String schemaStr) throws SQLException {
+  public static List<Schema.Field> getSchemaFields(ResultSet resultSet, @Nullable String schemaStr)
+    throws SQLException {
+    Schema resultsetSchema = Schema.recordOf("resultset", getSchemaFields(resultSet));
+
     if (!Strings.isNullOrEmpty(schemaStr)) {
       try {
         Schema schema = Schema.parseJson(schemaStr);
+        for (Schema.Field field : schema.getFields()) {
+          Schema.Field resultsetField = resultsetSchema.getField(field.getName());
+          if (resultsetField == null) {
+            throw new IllegalArgumentException(String.format("Schema field %s is not present in input record",
+                                                             field.getName()));
+          }
+          Schema resultsetFieldSchema = resultsetField.getSchema().isNullable() ?
+            resultsetField.getSchema().getNonNullable() : resultsetField.getSchema();
+          Schema simpleSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable() : field.getSchema();
+
+          if (!resultsetFieldSchema.equals(simpleSchema)) {
+            throw new IllegalArgumentException(String.format("Schema field %s has type %s but in input record found " +
+                                                               "type %s ",
+                                                             field.getName(), simpleSchema.getType(),
+                                                             resultsetFieldSchema.getType()));
+          }
+        }
         return schema.getFields();
       } catch (IOException e) {
-        LOG.error("Can not covert schema string {} to Schema", schemaStr);
+        throw new IllegalArgumentException(String.format("Can not covert schema string %s to Schema", schemaStr), e);
```

Comment:
its a little cleaner to just have the parseJson in this try-catch. Though nothing else in the try can throw an IOException, if somebody adds some logic later that does throw IOException, we'll end up with a misleading error message.